### ISSUE TITLE
contiguous() not work for rank 1 length 1 tensor.

### ISSUE THREAD

**Issue by hzhangxyz** [pytorch/pytorch#82610]
### 🐛 Describe the bug

When I try `torch.tensor([1.+2.j, 3.+4.j]).real.contiguous().stride()`, it returns `(1,)` as expected.

But if the length is only 1, the stride is not 1 but 2, that is to say `torch.tensor([1.+2.j]).real.contiguous().stride()` gives `(2,)`

Although stride of length 1 tensor has no effect. But when I try to use mpi4py to bcast/allreduce tensor, it will throw an error because of wrong stride.

### Versions

Collecting environment information...
PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.1.0
Clang version: 14.0.6
CMake version: version 3.23.3
Libc version: glibc-2.35

Python version: 3.10.5 (main, Jun  6 2022, 18:49:26) [GCC 12.1.0] (64-bit runtime)
Python platform: Linux-5.18.12-arch1-1-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.961
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.1
[pip3] torch==1.12.0
[conda] Could not collect

cc @ezyang @anjali411 @dylanbespalko @mruberry @Lezcano @nikitaved

**Comment by Lezcano** [comment]
A tensor of length 1 is always contiguous by definition, isn't it? All its elements (1) are stored in contiguous memory for trivial reasons.

**Comment by ezyang** [comment]
Yeah, the problem is that you have degrees of freedom for size 1 (and 0) dims, and PyTorch will accept all of them as contiguous. It's relatively easy to write a helper function that does some sort of strict contiguity, but we won't do it out of the box.

**Comment by Lezcano** [comment]
Another edge case that may bit you is the following:
Consider a tensor with arbitrary shape where one of the dimensions is equal to 1. In that case you can have the stride of that dimension to be zero, and that will also be a valid contiguous stride.

As Ed mentioned, this is a feature, not a bug, and may be worked around via a short custom function that clones the input if any of these edge cases is present.

Closing.

**Comment by dalcinl** [comment]
@Lezcano May I ask for your reconsideration? The behavior presented in the description is quite unexpected. For any dim=1 contiguous array, the stride should be 1 (except for len=0, where it would be best if stride=0, to follow NumPy rules). It is very hard to sell that for len>1 the stride is 1, but for len=1 the stride is 2. From general expression for multidimensional indexing, when applied to len=1 arrays, the i-th item is at address `(baseptr + i * stride)`.  The only reason the `len = 1` and `stride = 2` case would work properly is because the only valid index is `i = 0`, so any stride value would do. I grant you this is not a bug, but I hope you grant me my observation is fair and the current behavior seems messy/untidy.

**Comment by Lezcano** [comment]
You are completely right @dalcinl. The issue here is that, as @ezyang mentioned, for almost any given shape, there is exactly one stride that makes that shape contiguous. For example, for `shape = (4, 3, 2)`, the only strides that make it contiguous are `(6, 2, 1)`. We may call these "canonical strides" for that given shape. The canonical strides for a given shape may be computed as described in https://github.com/pytorch/pytorch/blob/master/c10/util/strides.h

Now, the issue is that there are some shapes that admit more strides (other than the canonical) that make them contiguous. These are shapes that have a dimension of length `0` or `1`. The posible strides for these dimensions are (examples below):
- A dimension of length `1` may have a stride equal to anything (!!!) and it will be contiguous.
- An empty tensor (any tensor that one of its dimensions has length zero) will be contiguous.

The algorithm is defined at https://github.com/pytorch/pytorch/blob/e63e97c59df09b54ac32160533dfc1046673142e/c10/core/TensorImpl.cpp#L234-L252

A few examples of the above to show that we may reach these states via valid PyTorch (and NumPy for the matter) constructs:
```python
import torch

# Example of 1D empty tensor with arbitrary stride
a = torch.rand(16)
b = a[::3]
print(b.stride())         # (3,)
c = b[0:0]
print(c.shape)            # (0,)
print(c.stride())         # (3,)
print(c.is_contiguous())  # True

# Example of 1D tensor with a dim of 1 and arbitrary stride
d = b[0:1]
print(d.shape)            # (1,)
print(d.stride())         # (3,)
print(d.is_contiguous())  # True

# Example of nD empty tensor with arbitrary stride
a = torch.rand(2, 16, 3)
b = a[:, ::3, :]
c = b[:, 0:0, :]
print(c.shape)            # (2, 0, 3)
print(c.stride())         # (48, 9, 1)
print(c.is_contiguous())  # True

# Example of nD tensor with a dim of 1 and stride of zero
a = torch.rand(2, 3)
b = a.unsqueeze(1)
c = b.expand(2, 2, 3)
d = c[:, 0:1, :]
print(d.shape)            # (2, 1, 3)
print(d.stride())         # (3, 0, 1)
print(d.is_contiguous())  # True

# Example of nD tensor with a dim of 1 and arbitrary stride
a = torch.rand(2, 3)
b = a.unsqueeze(1)
c = b[:, 0:1:2, :]
print(c.shape)            # (2, 1, 3)
print(c.stride())         # (3, 6, 1)
print(c.is_contiguous())  # True
```

**Comment by dalcinl** [comment]
@Lezcano I understand your conundrum. I'm not asking for stride normalization everywhere, that would be too much work for very little return. However, please look at what NumPy does here [[link]](https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/buffer.c#L501-L506) to alleviate the burden on buffer consumers. 

Would it be feasible to normalize stride values exported via DLPack, just as NumPy does? 
See the normalization in action:

```python
>>> import numpy as np
>>> a = np.arange(16, dtype='B')[::3][0:1]
>>> a.strides
(3,)
>>> memoryview(a).strides
(1,)
```

**Comment by Lezcano** [comment]
@rgommers may be more suited to handle this one

**Comment by rgommers** [comment]
> However, please look at what NumPy does here [[link]](https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/buffer.c#L501-L506) to alleviate the burden on buffer consumers.

I'll add the link to NumPy's DLPack implementation, which also ensures contiguous strides: https://github.com/numpy/numpy/blob/7c8e1134ae86f8a8c002e1068337bec63ddf7f0d/numpy/core/src/multiarray/dlpack.c#L138-L144

That is not completely canonical, but it is guaranteed to be a multiple of itemsize.

Let me also point out `NPY_RELAXED_STRIDES_DEBUG`, which is somewhat helpful for debugging issues with weird strides by scrambling them on purpose (from [link](https://numpy.org/devdocs/reference/arrays.ndarray.html?highlight=npy_relaxed_strides_debug)):

<img width="501" alt="image" src="https://user-images.githubusercontent.com/98330/182638059-1dd14e45-7408-4e62-8669-fe14281394dd.png">


> Would it be feasible to normalize stride values exported via DLPack, just as NumPy does?

That seems reasonable to me. Just to confirm @dalcinl: it's only about DLPack here, and not anything else (IIRC PyTorch doesn't export via the buffer protocol, but it does provide `__array__` and `__cuda_array_interface__`, and perhaps there's more)?

**Comment by dalcinl** [comment]
> Just to confirm @dalcinl: it's only about DLPack here

Yes.

> but it does provide `__array__` and `__cuda_array_interface__`

Well, mpi4py also supports `__cuda_array_interface__`. However, if DLPack is available, it takes precedence. So fixing `__cuda_array_interface__` is not a priority for mpi4py users.

**Comment by dalcinl** [comment]
> > Would it be feasible to normalize stride values exported via DLPack, just as NumPy does?
> 
> That seems reasonable to me.

After looking at the NumPy DLPack implementation you linked, looks like they do not actually normalize strides when exporting via DLPack. It is only for the Python buffer interface that they perform stride normalization.
